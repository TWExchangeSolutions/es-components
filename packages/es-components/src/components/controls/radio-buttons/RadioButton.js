import React from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeProvider, withTheme } from 'styled-components';
import { noop } from 'lodash';
import viaTheme from 'es-components-via-theme';

import Label from '../Label';
import getRadioFillVariables from './radio-fill-variables';

function radioFill(color) {
  return `
    background-color: ${color};
    border-radius: 100%;
    content: '';
    display: block;
    height: 13px;
    left: 3px;
    position: absolute;
    width: 13px;
    top: 3px;
    transition: background 0.25s linear;
  `;
}

const RadioLabel = styled(Label)`
  color: ${props => (props.error ? props.theme.colors.danger : 'inherit')};
  display: ${props => (props.inline ? 'inline-flex' : 'inherit')};
  font-size: ${props => props.theme.sizes.baseFontSize}px;
  font-weight: normal;
  margin-right: ${props => (props.inline ? '20px' : 'initial')};
  min-height: 25px;
  position: relative;
  padding: 5px 0;
  text-transform: none;

  &:hover .radio-fill:before {
    ${props => radioFill(props.hoverFillColor)};
  }
`;

const RadioInput = styled.input`
  opacity: 0;
  position: absolute;

  &:focus ~ .radio-fill {
    box-shadow: 0 0 3px 3px ${props => props.theme.colors.inputFocus};
  }
`;

const RadioText = styled.span`
  align-self: center;
  margin-left: 30px;
`;

const RadioDisplay = styled.span`
  border: 3px solid ${props => props.borderColor};
  border-radius: 100%;
  box-sizing: border-box;
  display: block;
  height: 25px;
  position: absolute;
  width: 25px;

  &:before {
    ${props => radioFill(props.fill)};
  }
`;

export function RadioButton({
  optionText,
  name,
  checked = false,
  id,
  isDisabled = false,
  inline = true,
  onClick = noop,
  value,
  hasError = false,
  theme,
  ariaHide,
  ...radioProps
}) {
  const { hover, fill } = getRadioFillVariables(
    checked,
    isDisabled,
    hasError,
    theme.colors
  );
  const radioDisplayFill = checked ? fill : theme.colors.white;

  const labelProps = {
    checked,
    inline,
    disabled: isDisabled,
    htmlFor: id,
    hoverFillColor: hover,
    error: hasError
  };

  return (
    <ThemeProvider theme={theme}>
      <RadioLabel {...labelProps}>
        <RadioInput
          type="radio"
          name={name}
          id={id}
          onClick={onClick}
          value={value}
          disabled={isDisabled}
          {...radioProps}
        />
        <RadioDisplay
          className="radio-fill"
          borderColor={fill}
          fill={radioDisplayFill}
        />
        <RadioText aria-hidden={ariaHide}>{optionText}</RadioText>
      </RadioLabel>
    </ThemeProvider>
  );
}

RadioButton.propTypes = {
  optionText: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  checked: PropTypes.bool,
  id: PropTypes.string,
  isDisabled: PropTypes.bool,
  inline: PropTypes.bool,
  onClick: PropTypes.func,
  value: PropTypes.any,
  ariaHide: PropTypes.bool,
  hasError: PropTypes.bool,
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object
};

RadioButton.defaultProps = {
  theme: viaTheme
};

export default withTheme(RadioButton);
