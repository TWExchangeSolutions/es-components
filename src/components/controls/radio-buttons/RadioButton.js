import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { noop } from 'lodash';

import Label from '../Label';
import { colors, sizes } from '../../theme';

import getRadioFillVariables from './radio-fill-variables';

function radioFill(color) {
  return `
    background-color: ${color};
    border-radius: 100%;
    content: '';
    display: block;
    height: 13px;
    left: 3px;
    position: absolute;
    width: 13px;
    top: 3px;
    transition: background 0.25s linear;
  `;
}

const RadioLabel = styled(Label)`
  color: ${props => (props.error ? colors.danger : 'inherit')};
  display: ${props => (props.inline ? 'inline-flex' : 'inherit')};
  font-size: ${sizes.baseFontSize}px;
  font-weight: normal;
  margin-right: ${props => (props.inline ? '20px' : 'initial')};
  min-height: 25px;
  position: relative;
  padding: 5px 0;
  text-transform: none;

  &:hover .radio-fill:before { ${props => radioFill(props.hoverFillColor)} }
`;

const RadioInput = styled.input`
  opacity: 0;
  position: absolute;

  &:focus ~ .radio-fill {
    box-shadow: 0 0 3px 3px #83bffc
  }
`;

const RadioText = styled.span`
  align-self: center;
  margin-left: 30px;
`;

const RadioDisplay = styled.span`
  border: 3px solid ${props => props.borderColor};
  border-radius: 100%;
  box-sizing: border-box;
  display: block;
  height: 25px;
  position: absolute;
  width: 25px;

  &:before { ${props => radioFill(props.fill)} }
`;

function RadioButton({
  optionText,
  name,
  checked = false,
  id,
  disabled = false,
  inline = true,
  onChange = noop,
  value,
  isInErrorState = false,
  ...radioProps
}) {
  const { hover, fill } = getRadioFillVariables(
    checked,
    disabled,
    isInErrorState
  );
  const radioDisplayFill = checked ? fill : colors.white;

  const labelProps = {
    checked,
    inline,
    disabled,
    htmlFor: id,
    hoverFillColor: hover,
    error: isInErrorState
  };

  return (
    <RadioLabel {...labelProps}>
      <RadioInput
        type="radio"
        name={name}
        id={id}
        onChange={onChange}
        value={value}
        disabled={disabled}
        {...radioProps}
      />
      <RadioDisplay
        className="radio-fill"
        borderColor={fill}
        fill={radioDisplayFill}
      />
      <RadioText>{optionText}</RadioText>
    </RadioLabel>
  );
}

RadioButton.propTypes = {
  optionText: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  checked: PropTypes.bool,
  id: PropTypes.string,
  disabled: PropTypes.bool,
  inline: PropTypes.bool,
  onChange: PropTypes.func,
  value: PropTypes.any,
  isInErrorState: PropTypes.bool
};

export default RadioButton;
