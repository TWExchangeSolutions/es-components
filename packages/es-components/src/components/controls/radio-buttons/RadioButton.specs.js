/* eslint-env jest */
/* eslint-disable max-len */

import React from 'react';
import { renderWithTheme } from '../../../testing';
import viaTheme from 'es-components-via-theme';

import getRadioFillVariables from './radio-fill-variables';
import { RadioButton } from './RadioButton';

describe('RadioButton component', () => {
  it('renders as expected', () => {
    const tree = renderWithTheme(
      <RadioButton id="test" optionText="test" name="rad" theme={viaTheme} />
    );

    expect(tree).toMatchSnapshot();
  });
});

describe('getRadioFillVariables', () => {
  it('returns { fill: gray5, hover: white } when unchecked, disabled, and valid', () => {
    const checked = false;
    const disabled = true;
    const invalid = false;

    const { fill, hover } = getRadioFillVariables(
      checked,
      disabled,
      invalid,
      viaTheme.colors
    );

    expect(fill).toBe(viaTheme.colors.gray5);
    expect(hover).toBe(viaTheme.colors.white);
  });

  it('returns { fill: info } when checked, not disabled, and valid', () => {
    const checked = true;
    const disabled = false;
    const invalid = false;

    const { fill } = getRadioFillVariables(
      checked,
      disabled,
      invalid,
      viaTheme.colors
    );

    expect(fill).toBe(viaTheme.colors.info);
  });

  it('returns { fill: gray5 } when checked, disabled, and valid', () => {
    const checked = true;
    const disabled = true;
    const invalid = false;

    const { fill } = getRadioFillVariables(
      checked,
      disabled,
      invalid,
      viaTheme.colors
    );

    expect(fill).toBe(viaTheme.colors.gray5);
  });

  it('returns { fill: danger, hover: danger } when unchecked, not disabled, and invalid', () => {
    const checked = false;
    const disabled = false;
    const invalid = true;

    const { fill, hover } = getRadioFillVariables(
      checked,
      disabled,
      invalid,
      viaTheme.colors
    );

    expect(fill).toBe(viaTheme.colors.danger);
    expect(hover).toBe(viaTheme.colors.danger);
  });

  it('returns { fill: danger } when checked, not disabled, and invalid', () => {
    const checked = true;
    const disabled = false;
    const invalid = true;

    const { fill } = getRadioFillVariables(
      checked,
      disabled,
      invalid,
      viaTheme.colors
    );

    expect(fill).toBe(viaTheme.colors.danger);
  });

  it('returns { fill: gray8, hover: gray8 } when unchecked, not disabled, and valid', () => {
    const checked = false;
    const disabled = false;
    const invalid = false;

    const { fill, hover } = getRadioFillVariables(
      checked,
      disabled,
      invalid,
      viaTheme.colors
    );

    expect(fill).toBe(viaTheme.colors.gray8);
    expect(hover).toBe(viaTheme.colors.gray8);
  });
});
