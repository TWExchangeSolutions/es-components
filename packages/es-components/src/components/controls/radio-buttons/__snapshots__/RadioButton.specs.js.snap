// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton component renders as expected 1`] = `
<label
  checked={false}
  className="sc-bwzfXH kVIXAD sc-bdVaJa kHjvzh"
  disabled={false}
  htmlFor="test"
>
  <input
    className="sc-htpNat jeuMMj"
    disabled={false}
    id="test"
    name="rad"
    onClick={[Function]}
    type="radio"
    value={undefined}
  />
  <span
    className="radio-fill sc-ifAKCX fnveVG"
    fill="#fff"
  />
  <span
    aria-hidden={undefined}
    className="sc-bxivhb bRNnnP"
  >
    test
  </span>
</label>
`;
